File: yunhwan/프로그래머스 순위.py
def solution(n, results):
    answer = 0
    
    #1. 그래프 생성 및 초기화
    #2. 그래프 순회하며 dfs 실행
    #   도착점인 점들이 방문하지 않은 경우,
    #   arr추가 및 dfs 탐색
    #3. 그래프를 순회하면서 시작점인 경우, 도착점인 경우를 세주기
    #4. cnt배열의 개수가 ,n-1개인 경우 만족하는 경우 answer+=1 해주기

    
    g = [[] for _ in range(n+1)]
    v = [0]*(n+1)
    cnt = [0] * (n+1)
    
    #그래프 초기화
    for i in results:
        g[i[0]].append(i[1])


    
    # 도착점인 점들이 방문하지 않은 경우,
    # arr추가 및 dfs 탐색
    def dfs(start,arr):
        
        for i in g[start]:
            if v[i]==0:
                arr.append(i)
                v[i]=1
                dfs(i,arr)
            
        return arr


    # 그래프 순회하며 방문 리스트 초기화 및 dfs 탐색     
    for i in range(1,n+1):
        v = [0]*(n+1)
        g[i] = dfs(i,[])

    #그래프를 순회
    for i in range(1,n+1):
        #i가 시작점인 개수를 더해주기
        cnt[i]+=len(g[i])
        for j in g[i]:
            #도착점인 경우를 더해주기
            cnt[j]+=1
    
    #cnt 배열을 순회하며
    # n-1을 만족하는 경우, answer+=1 해주기
    for i in range(1,n+1):
        if cnt[i]==n-1:
            answer+=1
    

    return answer

File: jiWoo/프로그래머스/프로그래머스_순위.java
package jiWoo.프로그래머스;

import java.util.*;
public class 프로그래머스_순위 {
    static boolean[] visited;
    // 위너들이 이긴 사람(loser) 넣어놓은 리스트
    static ArrayList<Integer>[] winner;
    public int solution(int n, int[][] results) {
        int answer = 0;
        visited=new boolean[n+1];
        winner=new ArrayList[n+1];
        for(int i=0; i<=n; i++) //
            winner[i]=new ArrayList<>();
        // results에 있는 결과 넣어주기 result[i][0]: 승자 result[i][1] 패자
        int len=results.length;
        for(int i=0; i<len; i++){
            winner[results[i][0]].add(results[i][1]);
        }

        // DFS를 사용하여 승자에 패자 리스트 전부 채워주기
        for(int i=1; i<=n; i++){
            visited=new boolean[n+1];
            winner[i]=DFS(i,new ArrayList<Integer>());
        }

        // winner[i]에는 각 선수별 탐색할 수 있는 모든 패자들을 넣어주었다.
        // 정확하게 순위를 매길 수 있는 선수는 이긴 수 + 진 수 = n-1여야 한다.
        int [] cnt=new int[n+1];

        for(int i=1; i<=n; i++){
            // i 이긴 수 체크
            cnt[i]+=winner[i].size(); // i 의 패자리스트의 개수 더해주기
            // i의 패자리스트에 값-> j의 진 수 체크

            for(int j:winner[i]){
                cnt[j]++; // j는 i에게 졌으므로 cnt[j]++
            }
        }

        // cnt[i]=n-1일 때 정확하게 순위 매길 수 있다.
        for(int sum:cnt){
            if(sum==(n-1))answer++;
        }

        return answer;
    }

    public static ArrayList<Integer> DFS(int i, ArrayList<Integer> li){
        // 위너의 패자를 탐색하며 그 패자의 패자들을 넣어주기 위한 DFS
        for(int loser:winner[i]){

            if(!visited[loser]){
                visited[loser]=true;
                li.add(loser);
                DFS(loser,li);
            }
        }
        return li;
    }
}
